core/src/banking_stage/thread_aware_account_locks.rs

```diff
@@ -0,0 +1,692 @@
+use {
+    solana_sdk::pubkey::Pubkey,
+    std::{
+        collections::{hash_map::Entry, HashMap},
+        fmt::{Debug, Display},
+        ops::{BitAndAssign, Sub},
+    },
+};
+
+pub const MAX_THREADS: usize = u64::BITS as usize;
+
+/// Identifier for a thread
+pub type ThreadId = usize; // 0..MAX_THREADS-1
+
+/// A bit-set of threads an account is scheduled or can be scheduled for.
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct ThreadSet(u64);
+
+/// Thread-aware account locks which allows for scheduling on threads
+/// that already hold locks on the account. This is useful for allowing
+/// queued transactions to be scheduled on a thread while the transaction
+/// is still being executed on the thread, up to a queue limit.
+pub struct ThreadAwareAccountLocks {
+    /// Number of threads.
+    num_threads: usize, // 0..MAX_THREADS
+    /// Limit on the number of sequentially-queued transactions per account.
+    sequential_queue_limit: u32,
+    /// Write locks - only on thread can hold a write lock at a time.
+    /// Contains how many write locks are held by the thread.
+    write_locks: HashMap<Pubkey, (ThreadId, u32)>,
+    /// Read locks - multiple threads can hold a read lock at a time.
+    /// Contains thread-set for easily checking which threads are scheduled.
+    /// Contains how many read locks are held by each thread.
+    read_locks: HashMap<Pubkey, (ThreadSet, [u32; MAX_THREADS])>,
```

Comment:
also how about introducing `type LockCount = u32;`? it would help readers to notice those `u32`s are in the same dimension of units.